internal/interface/httpapi/api.go

```diff
@@ -216,15 +222,150 @@ func (a *Api) deleteLink(writer http.ResponseWriter, request *http.Request) {
 	writer.WriteHeader(http.StatusOK)
 }
 
+func urlChecker(ctx context.Context, in <-chan string, out chan<- bool, stop <-chan struct{}) error {
+	for {
+		select {
+		case <-stop:
+			return nil
+		case <-ctx.Done():
+			return ctx.Err()
+		case u, ok := <-in:
+			if !ok {
+				return nil
+			}
+			flag, err := checkUrl(ctx, u)
+			if err != nil {
+				return err
+			}
+			out <- flag
+		}
+	}
+}
+
+func checkUrl(ctx context.Context, urlStr string) (bool, error) {
+	c := http.Client{
+		Timeout: 5 * time.Second,
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, urlStr, nil)
+	if err != nil {
+		return false, err
+	}
+	resp, err := c.Do(req)
+	if err != nil {
+		if e, ok := err.(*url.Error); ok {
+			if e.Timeout() || e.Temporary() {
+				return false, nil
+			} else {
+				return false, err
+			}
+		} else {
+			return false, err
+		}
+	}
+	if resp != nil {
+		defer resp.Body.Close()
+		if resp.StatusCode == http.StatusNotFound {
+			return false, nil
+		} else {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+const (
+	urlsConcurrency = 4
+)
+
+func getLinksAvailability(urls []string) ([]bool, error) {
```

Comment:
So, you are making some pipeline running on API call? 
And you are handling OS signal from the context of API call handler? Am I right?

Why do you do this in the context of user request in the first place? It could be done completely independently.

Reply:
I'll try to clarify my point. It is very disturbing to see OS signal handling at a random place. What are you trying to achieve here? Will somebody send a signal to cancel exactly this API call? Will you add signal handling in each of API calls?
If you want to have a possibility to turn down pipeline use more idiomatic and clean way. Place OS signals handling near the entry point of you program and use `context` package to propagate signals.

Reply:
ok, i create context for api and move signal handling here